### PR TITLE
Add optimizer

### PR DESCRIPTION
**Pull request by kevinthesun**
1. Add optimizer API.
2. Add SGD and Adam optimizers.
3. Add learning rate scheduler API.

Currently the state variables of optimizer, such as momentum and number of updates, need to be explicitly initialized when building the library. it might be able to improve this process.

### CODE REVIEW

File: src/top/tensor/state_op.cc [hunk 1]
```diff
@@ -0,0 +1,36 @@
+/*!
+ *  Copyright (c) 2017 by Contributors
+ * \file matrix_op.cc
```
Comment (ZihengJiang) [inline src/top/tensor/state_op.cc]: change comment

**Comment by kevinthesun** [comment]
@tqchen  @ZihengJiang  Since we are moving operators to c++, I think optimizer should also be implemented in c++ so that it can be directly registered by other languages. What's your opinion?

**Comment by ZihengJiang** [comment]
@kevinthesun no hurry for this

**Comment by kevinthesun** [comment]
OK. I'll try to merge https://github.com/dmlc/nnvm/pull/349 into this PR so that we can have end to end example of how to do training in nnvm.

**Comment by nhynes** [comment]
@kevinthesun any movement on this? This is something I'd be interested in having soon, so please let me know how I can help!

File: src/top/tensor/state_op.cc [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+#include <nnvm/node.h>
+#include <nnvm/op_attr_types.h>
+#include <nnvm/top/tensor.h>
+#include "../op_common.h"
+#include "../elemwise_op_common.h"
+
+namespace nnvm {
+namespace top {
+
+NNVM_REGISTER_OP(assign)
```
Comment (tqchen) [inline src/top/tensor/state_op.cc]: OK, stateful ops requires consideration and the current support may not be perfect. Let us first change the name to ```_assign``` for now

File: tests/python/unittest/test_optimizer.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    scheduler = lr_scheduler.FactorScheduler(step=5, factor=0.6)
+    opt = optimizer.Adam(lr_scheduler=scheduler,
+                         rescale_grad=0.2, clip_gradient=1.0, wd=0.2)
+    opt_sym = opt.minimize(y)
+    assert len(opt_sym.list_output_names()) == 1
+    g = nnvm.graph.create(opt_sym).ir()
+    assert 'Adam_t' in g
+    assert 'Adam_m0' in g
+    assert 'Adam_v0' in g
+
```
Comment (tqchen) [inline tests/python/unittest/test_optimizer.py]: please add a test-case for stateful ops with simple assign. We might need to update graph runtime to add get_input, as the states belongs to input variable

Comment (kevinthesun) [reply tests/python/unittest/test_optimizer.py]: Yes. We need to consider such kind of in-place manner operators. Currently it looks like in tvm there is no such function primitives.

Comment (kevinthesun) [reply tests/python/unittest/test_optimizer.py]: In tvm, assign can be prototyped as writing tensor values directly to assigned tensor memory. Do we have a elegant way to implement this in tvm?

Comment (kevinthesun) [reply tests/python/unittest/test_optimizer.py]: Or in short term, I can go ahead to address comments and make nnvm training graph complete. We can consider tvm implementation later.

Comment (tqchen) [reply tests/python/unittest/test_optimizer.py]: We can pass in arguments for both input and output in TVM, for elemwise operation they can be the same.

For example
```python
x = tvm.placeholder(shape)
y = tvm.placeholder(shape)
z = topi.elemwise_add(x, y)

s = tvm.create_schedule(z.op)
f = tvm.build(s, [x, y, z])

data = tvm.nd.array(shape)
rhs = tvm.nd.array(shape)
# data = data + rhs
f(data, rhs, data)
```

Comment (kevinthesun) [reply tests/python/unittest/test_optimizer.py]: While we can have a built function for assign in tvm, we still need to register in nnvm. However, in nnvm register_compute function, we can't build a tvm function and utilize tvm.nd.array interface to move tensor values. Is my understanding correct?

Comment (tqchen) [reply tests/python/unittest/test_optimizer.py]: That is true, a special registration mechanism is needed to tell NNVM compiler to generate the operator that mutates its input. Need to update the NNVM compiler's build logic to support this

Comment (kevinthesun) [reply tests/python/unittest/test_optimizer.py]: Is it possible to directly manipulate tvm tensor internal data in FTVMCompute function? 
```
.set_attr<FTVMCompute>(
  "FTVMCompute", [](const NodeAttrs& attrs,
                    const Array<Tensor>& inputs,
                    const Array<Tensor>& out_info) {
      rtv = Array<Tensor>{ topi::identity(inputs[1]) };
      // Some manipulation to change internal data structure to make inputs[0] 'points' to rtv
      inputs[0]...
      return rtv;
})
```

**Comment by tqchen** [comment]
https://github.com/dmlc/nnvm/pull/389 partially resolves the assign problem, please rebase after that is merged

**Comment by kevinthesun** [comment]
Add initializing variables during building graph. Some tensor operators are the same as https://github.com/dmlc/nnvm/pull/401. 

Add test cases for SGD and Adam. These test cases can pass locally on mac OS and ubuntu. However, they somehow failed in CI .

File: docs/top.rst [hunk 4]
```diff
@@ -58,6 +57,10 @@ This level enables fully connected multi-layer perceptron.
    nnvm.symbol.pad
    nnvm.symbol.block_grad
    nnvm.symbol.indicator
+   nnvm.symbol.clip
+   nnvm.symbol.assign
```
Comment (tqchen) [inline docs/top.rst]: assign is experimental and should not be part of the doc

File: docs/top.rst [hunk 5]
```diff
@@ -58,6 +57,10 @@ This level enables fully connected multi-layer perceptron.
    nnvm.symbol.pad
    nnvm.symbol.block_grad
    nnvm.symbol.indicator
```
Comment (tqchen) [inline docs/top.rst]: indicator should not be part of level 1

**Comment by tqchen** [comment]
Let us move the newly added operators' support level to 4 because they are not necessary needed to support common convnets. This can be merged after the change

**Comment by kevinthesun** [comment]
Moved all newly added operators to level4.

**Comment by nhynes** [comment]
You forgot to add https://github.com/dmlc/nnvm/pull/349 before merging this.

**Comment by kevinthesun** [comment]
@nhynes All those operators except matmul were added. Matmul will be fixed by https://github.com/dmlc/nnvm/pull/396

**Comment by nhynes** [comment]
ha, you just beat me to submitting a pr. I'll close that now. thanks!
